doc/port-fec/fec-auto-determination.md

```diff
@@ -0,0 +1,185 @@
+# SONiC FEC Auto Determination Design #
+
+## Table of Content
+
+- [Revision](#revision)
+- [Scope](#scope)
+- [Definitions/Abbreviations](#definitions/abbreviations)
+- [Overview](#overview)
+- [High-Level Design](#high-level-design)
+- [API design](#api-design)
+- [Common Rule for FEC Determination](#common-rule-for-fec-determination)
+    - [Table 1: FEC Mapping Based on Optics Type](#table-1-fec-mapping-based-on-optics-type)
+    - [Table 2: FEC Mapping Based on Lane Speed and Number of Lanes](#table-2-fec-mapping-based-on-lane-speed-and-number-of-lanes)
+- [Diagram For Different Use Cases](#diagram-for-different-use-cases)
+- [Dependency](#dependency)
+
+### Revision
+
+ | Rev |     Date    |       Author        | Change Description                         |
+ |:---:|:-----------:|:-------------------:|--------------------------------------------|
+ | 0.1 |             |     Shyam Kumar, Longyin Huang   | Initial version                            |
+
+### Scope
+This document is the design document for FEC auto-determination feature on SONiC.
+
+### Definitions/Abbreviations
+| **Term**       | **Definition**                                   |
+| -------------- | ------------------------------------------------ |
+| FEC            | Forward Error Correction                         |
+| DPB            | Dynamic Port Breakout                            |
+
+### Overview
+
+FEC mode is a critical configuration for a port, which needs to be configured properly for the port to come up.
+
+There are below scenarios that can end up with wrong FEC mode:
+1. In DPB(Dynamic Port Breakout) case, today's DPB CLI doesn't generate FEC config for newly created ports, FEC mode is default to `none` at [SAI/SDK](https://github.com/opencomputeproject/SAI/blob/a94bbbe43242a4d9e1a4d9f70780ea9251127f5d/inc/saiport.h#L1012) layer.
+2. In non-DPB case,
+    - Some platforms have no FEC configured in CONFIG_DB by default. The FEC mode can be either default to `none` at SAI/SDK layer or manually configured by user who might not have enough domain knowledge.
+    - Some platforms have default FEC defined in `port_config.ini`, which however might not be suitable for the specific port/optics on the system.
+
+The feature in this document is to address the issue in both of above scenarios in a common platform-independent way, since the rule to determine FEC for a given port/optics is common for all platforms.
+
+### High-Level Design
+
+- Add `determine-fec` module which can determine FEC mode based on common rule for a given port with a given optics.
+    - This module provides a `determine_fec` API which can be invoked by below entities.
+- Enhance today's DPB CLI to automatically determine and configure FEC in CONFIG_DB for newly created ports, based on determine-fec module.
+- Add a user-triggered CLI `fec-auto-correct` to automatically determine and configure FEC in CONFIG_DB for existing ports, based on determine-fec module.
```

Comment:
How about having this integrated into pmon#xcvrd to automate the entire process?

i.e.
1. Introduce a platform-independent common rule file as a .json for pmon#xcvrd (e.g. sonic-xcvrd/xcvrd/data/media_autoconf.json)
2. pmon#xcvrd loads the common rules at startup and have it override if a platform-specific media_autoconf.json is present in its device folder.
3. Upon SFP insertion, automatically determine the 'Recommended FEC type' based on the transceiver type detected on a port, and then generate a per port "auto_fec=rs|fc|none" to STATE_DB (and APPL_DB)
4. Add "auto" to the admin FEC type, which means the per port FEC type in CONFIG_DB/APPL_DB will be "fec=auto|rs|fc|none"
5. When fec=auto, the OA should configure the per port FEC type based on "auto_fec" from pmon#xcvrd, and fallbacks to 'none' for NRZ port, 'rs' for PAM4 ports,  when 'auto_fec' is not present.

Reply:
One more thing, the *hint* from pmon#xcvrd could also be useful when we get started working on the CL73 AutoNeg integration, we'll need to adjust the FEC type advertisement based on the transceiver detected on the port

Reply:
Yea, we already mentioned xcvrd integration in this HLD:
https://github.com/sonic-net/SONiC/blob/35b5585d31bf29c83587fc4a5066c82d871e0c8a/doc/port-fec/fec-auto-determination.md?plain=1#L176
The plan is to treat this as future enhancement, which can be an incremental update(via separate PR) on this HLD. And more detail will be added at that time.

Thanks for the suggestion on the implementation detail, which will be taken into consideration.